File: src/update.ts
// The in-app updater: the check against the channel's feed, the download +
// install the user asked for, the footer mark and the App Update dialog's
// buttons (VersionDialog.vue reads the state). Nothing is downloaded or
// installed without a click. The work happens in the backend (`update.rs`:
// `check_update`, `install_update` with `update-progress` events), because
// only that side can pick the channel's feed per check. App.vue loads this
// module on demand, only when the backend says the install is one the
// updater can replace (`SystemInfo.updater`: the Windows installer and the
// AppImage; the bare executable and deb / rpm never call any of this).
import { markRaw, reactive, type Component } from "vue";
import { invoke } from "@tauri-apps/api/core";
import { listen } from "@tauri-apps/api/event";
import UpdateMark from "./components/UpdateMark.vue";
import type { ConfirmButton } from "./components/ConfirmDialog.vue";
import type { UpdateChannel } from "./types";

// idle: no check yet; current: up to date; available: an update waits for
// the user's click; downloading / installing: after it; error: the last
// check or install failed (`error` says which).
export type UpdateState = "idle" | "checking" | "current" | "available" | "downloading" | "installing" | "error";

// `check_update`'s answer for a found update.
export interface UpdateInfo {
  version: string;
  // Release date as `YYYY-MM-DD`, empty when the feed has none.
  date: string;
  // Release notes, as the feed carries them.
  notes: string;
}

// `update-progress` event payload.
type Progress = { kind: "download"; downloaded: number; total: number | null } | { kind: "installing" };

export interface Updater {
  state: UpdateState;
  info: UpdateInfo | null;
  // Dev builds only (`tauri dev` has no updater): a simulated updater; the
  // dialog then carries a toggle that fakes an available update, so the GUI
  // parts can be looked at. null for a real updater.
  simulated: boolean | null;
  // Download progress 0..1; null until the size is known.
  progress: number | null;
  error: string;
  // The footer's mark while an update waits (UpdateMark.vue).
  mark: Component;
  // Check the channel's feed; true when an update is available.
  check(): Promise<boolean>;
  // Download and install the available update; the app restarts after.
  install(): Promise<void>;
  // The version dialog's update buttons for the current state.
  buttons(): ConfirmButton[];
  // Simulated updater: fake an available update, or reset.
  simulate(on: boolean): void;
}

// `channel`: the Settings choice, read at every check. `simulate`: a
// dev-only stand-in (never talks to the backend) instead of the real thing;
// a release build drops the simulation code.
export function createUpdater(channel: () => UpdateChannel, simulate = false): Updater {
  let fakeTimer: ReturnType<typeof setInterval> | null = null;

  const u: Updater = reactive({
    state: "idle" as UpdateState,
    info: null,
    simulated: import.meta.env.DEV && simulate ? false : null,
    progress: null,
    error: "",
    mark: markRaw(UpdateMark),

    async check() {
      if (import.meta.env.DEV && u.simulated !== null) {
        // A fake check: "Checking…" for a moment, then what the toggle says.
        if (u.state !== "idle" && u.state !== "current" && u.state !== "available") return false;
        u.state = "checking";
        await new Promise((r) => setTimeout(r, 900));
        u.state = u.simulated ? "available" : "current";
        return u.state === "available";
      }
      if (u.state === "checking" || u.state === "downloading" || u.state === "installing") return u.state !== "checking";
      u.state = "checking";
      u.error = "";
      try {
        const found = await invoke<UpdateInfo | null>("check_update", { channel: channel() });
        if (!found) {
          u.info = null;
          u.state = "current";
          return false;
        }
        u.info = found;
        u.state = "available";
        return true;
      } catch (e) {
        u.error = `Check failed: ${e}`;
        u.state = "error";
        console.warn("update check failed", e);
        return false;
      }
    },

    async install() {
      if (u.state !== "available") return;
      if (import.meta.env.DEV && u.simulated !== null) {
        // A fake download: 2 % every 60 ms, then "installing" for good.
        u.state = "downloading";
        u.progress = 0;
        fakeTimer = setInterval(() => {
          u.progress = Math.min(1, (u.progress ?? 0) + 0.02);
          if (u.progress >= 1) {
            if (fakeTimer) clearInterval(fakeTimer);
            fakeTimer = null;
            u.state = "installing";
          }
        }, 60);
        return;
      }
      u.state = "downloading";
      u.progress = null;
      u.error = "";
      const unlisten = await listen<Progress>("update-progress", (e) => {
        const p = e.payload;
        if (p.kind === "download") u.progress = p.total ? Math.min(1, p.downloaded / p.total) : null;
        else u.state = "installing";
      });
      try {
        // Returns only on failure: success ends in a restart.
        await invoke("install_update");
      } catch (e) {
        u.error = `Install failed: ${e}`;
        u.state = "error";
        console.error("update install failed", e);
      } finally {
        unlisten();
      }
    },

    buttons() {
      const busy = u.state === "checking" || u.state === "downloading" || u.state === "installing";
      // Check Update until one is found; Install from then on.
      if (u.state === "available" || u.state === "downloading" || u.state === "installing") {
        return [{ label: "Install", kind: "primary", value: "install", disabled: busy }];
      }
      return [{ label: "Check Update", kind: "outline", value: "check", disabled: busy }];
    },

    simulate(on) {
      if (!import.meta.env.DEV || u.simulated === null) return;
      if (fakeTimer) clearInterval(fakeTimer);
      fakeTimer = null;
      u.simulated = on;
      u.progress = null;
      u.error = "";
      if (on) {
        u.info = {
          version: "9.9.9",
          date: new Date().toISOString().slice(0, 10),
          notes: "Simulated release notes.\n\n- A first change\n- A second change",
        };
        u.state = "available";
      } else {
        u.info = null;
        u.state = "idle";
      }
    },
  });
  return u;
}

File: src-tauri/src/update.rs
//! The in-app updater's backend: the feed per channel, the check, the
//! download + install with progress events. The updater plugin's own IPC
//! commands are not used — only this side can pick the endpoint per check,
//! and the channel is a setting (`Config::update_channel`).
//!
//! Feeds: the stable channel reads GitHub's `latest` release (never a
//! pre-release); the prerelease channel reads the rolling
//! `prerelease-version` release, whose `latest.json` CI points at the newest
//! published release, pre-release or stable (`.github/workflows/release.yml`).
//! A pre-release is the same binary and version number as the stable it may
//! become: publishing it as a full release promotes it, no rebuild.
//!
//! An install without the updater (bare executable, deb / rpm) only checks:
//! the feed's version against its own, the frontend links the project page.

use std::sync::Mutex;

use log::{info, warn};
use serde::Serialize;
use tauri::{AppHandle, Emitter, State};
use tauri_plugin_updater::{Update, UpdaterExt};

use crate::config::UpdateChannel;

pub const STABLE_FEED: &str = "https://github.com/w00zla/BindSight/releases/latest/download/latest.json";
pub const PRERELEASE_FEED: &str = "https://github.com/w00zla/BindSight/releases/download/prerelease-version/latest.json";
const GITHUB_REPO: &str = "w00zla/BindSight";

/// The feed a channel reads.
pub fn feed_url(channel: UpdateChannel) -> &'static str {
    match channel {
        UpdateChannel::Stable => STABLE_FEED,
        UpdateChannel::Prerelease => PRERELEASE_FEED,
    }
}

/// The GitHub release body (the changelog) for a version's `v<version>` tag,
/// as plain text. Best-effort: any failure yields an empty string, because
/// release notes are cosmetic and must never fail the update check. A
/// pre-release lives under its real version tag too, so this works on both
/// channels.
async fn fetch_release_notes(version: &str) -> String {
    let url = format!("https://api.github.com/repos/{GITHUB_REPO}/releases/tags/v{version}");
    match release_body(&url).await {
        Ok(body) => body,
        Err(e) => {
            warn!("release notes fetch ({url}) failed: {e}");
            String::new()
        }
    }
}

async fn release_body(url: &str) -> Result<String, String> {
    let json = get_json(url).await?;
    Ok(json.get("body").and_then(|b| b.as_str()).unwrap_or_default().trim().to_string())
}

/// GET a JSON document from GitHub.
async fn get_json(url: &str) -> Result<serde_json::Value, String> {
    let client = reqwest::Client::builder()
        // GitHub rejects requests without a User-Agent.
        .user_agent(concat!("BindSight/", env!("CARGO_PKG_VERSION")))
        // Never let a slow server stall the update check.
        .timeout(std::time::Duration::from_secs(10))
        .build()
        .map_err(|e| e.to_string())?;
    let resp = client
        .get(url)
        .header("Accept", "application/vnd.github+json")
        .send()
        .await
        .map_err(|e| e.to_string())?;
    if !resp.status().is_success() {
        return Err(format!("HTTP {}", resp.status()));
    }
    let text = resp.text().await.map_err(|e| e.to_string())?;
    serde_json::from_str(&text).map_err(|e| e.to_string())
}

/// Whether the feed's version is newer than the running one.
fn is_newer(feed: &str, current: &str) -> Result<bool, String> {
    let parse = |v: &str| semver::Version::parse(v.trim().trim_start_matches('v')).map_err(|e| format!("version {v:?}: {e}"));
    Ok(parse(feed)? > parse(current)?)
}

/// An install without the updater (bare executable, deb / rpm): read the
/// channel's feed by hand and compare versions; the frontend links the
/// project page, nothing is downloaded.
async fn check_feed(url: &str) -> Result<Option<UpdateInfo>, String> {
    let feed = get_json(url).await?;
    let version = feed.get("version").and_then(|v| v.as_str()).ok_or("feed without a version")?;
    if !is_newer(version, env!("CARGO_PKG_VERSION"))? {
        return Ok(None);
    }
    let version = version.trim().trim_start_matches('v').to_string();
    let gh = fetch_release_notes(&version).await;
    let str_of = |k: &str| feed.get(k).and_then(|v| v.as_str()).unwrap_or_default().to_string();
    Ok(Some(UpdateInfo {
        date: str_of("pub_date").get(..10).unwrap_or_default().to_string(),
        notes: if gh.is_empty() { str_of("notes") } else { gh },
        version,
    }))
}

/// The update the last check found, kept for `install_update`.
#[derive(Default)]
pub struct UpdateState(Mutex<Option<Update>>);

/// What the frontend shows about a found update.
#[derive(Debug, Serialize)]
pub struct UpdateInfo {
    pub version: String,
    /// `YYYY-MM-DD`, empty when the feed has none.
    pub date: String,
    pub notes: String,
}

/// `update-progress` event payload.
#[derive(Debug, Clone, Serialize)]
#[serde(tag = "kind", rename_all = "lowercase")]
pub enum Progress {
    Download { downloaded: u64, total: Option<u64> },
    Installing,
}

/// Check the channel's feed; the found update (if any) is kept for
/// `install_update`.
#[tauri::command]
pub async fn check_update(
    channel: UpdateChannel,
    app: AppHandle,
    state: State<'_, UpdateState>,
) -> Result<Option<UpdateInfo>, String> {
    let url = feed_url(channel);
    if !crate::updater_available() {
        let info = check_feed(url).await.map_err(|e| {
            warn!("update check ({channel:?}) failed: {e}");
            e
        })?;
        match &info {
            Some(i) => info!("update check ({channel:?}): v{} available", i.version),
            None => info!("update check ({channel:?}): up to date"),
        }
        return Ok(info);
    }
    let updater = app
        .updater_builder()
        .endpoints(vec![url.parse().map_err(|e| format!("{url}: {e}"))?])
        .map_err(|e| e.to_string())?
        .build()
        .map_err(|e| e.to_string())?;
    let found = updater.check().await.map_err(|e| {
        warn!("update check ({channel:?}) failed: {e}");
        e.to_string()
    })?;
    let info = if let Some(u) = found.as_ref() {
        // Prefer the GitHub release's changelog; fall back to the feed's own
        // `notes` when it cannot be fetched.
        let gh = fetch_release_notes(&u.version).await;
        Some(UpdateInfo {
            version: u.version.clone(),
            date: u.date.map(|d| d.date().to_string()).unwrap_or_default(),
            notes: if gh.is_empty() { u.body.clone().unwrap_or_default() } else { gh },
        })
    } else {
        None
    };
    match &info {
        Some(i) => info!("update check ({channel:?}): v{} available", i.version),
        None => info!("update check ({channel:?}): up to date"),
    }
    *state.0.lock().unwrap() = found;
    Ok(info)
}

/// Download and install the update the last check found, reporting
/// `update-progress`, then restart. On Windows the installer takes over and
/// ends the process itself.
#[tauri::command]
pub async fn install_update(app: AppHandle, state: State<'_, UpdateState>) -> Result<(), String> {
    if !crate::updater_available() {
        return Err("no updater in this install".into());
    }
    let update = state.0.lock().unwrap().clone().ok_or("no update checked")?;
    info!("update v{} downloading", update.version);
    let mut downloaded: u64 = 0;
    let on_chunk = app.clone();
    let on_done = app.clone();
    update
        .download_and_install(
            |chunk, total| {
                downloaded += chunk as u64;
                let _ = on_chunk.emit("update-progress", Progress::Download { downloaded, total });
            },
            || {
                let _ = on_done.emit("update-progress", Progress::Installing);
            },
        )
        .await
        .map_err(|e| {
            warn!("update install failed: {e}");
            e.to_string()
        })?;
    info!("update v{} installed, restarting", update.version);
    app.restart()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn feeds_per_channel() {
        assert!(feed_url(UpdateChannel::Stable).ends_with("/releases/latest/download/latest.json"));
        assert!(feed_url(UpdateChannel::Prerelease).ends_with("/releases/download/prerelease-version/latest.json"));
        assert_ne!(feed_url(UpdateChannel::Stable), feed_url(UpdateChannel::Prerelease));
        for url in [STABLE_FEED, PRERELEASE_FEED] {
            assert!(url.starts_with("https://"), "{url}");
        }
    }

    #[test]
    fn newer_versions() {
        assert_eq!(is_newer("0.17.0", "0.16.0"), Ok(true));
        assert_eq!(is_newer("v1.0.0", "0.16.9"), Ok(true));
        assert_eq!(is_newer("0.16.0", "0.16.0"), Ok(false));
        assert_eq!(is_newer("0.15.2", "0.16.0"), Ok(false));
        assert!(is_newer("latest", "0.16.0").is_err());
        assert!(is_newer("", "0.16.0").is_err());
    }

    #[test]
    fn progress_serializes_with_a_kind_tag() {
        let json = serde_json::to_value(Progress::Download { downloaded: 5, total: Some(10) }).unwrap();
        assert_eq!(json["kind"], "download");
        assert_eq!(json["downloaded"], 5);
        assert_eq!(serde_json::to_value(Progress::Installing).unwrap()["kind"], "installing");
    }
}
